confidence/confidence.go

```diff
@@ -0,0 +1,198 @@
+package confidence
+
+import (
+	"context"
+	"fmt"
+	"time"
+	"net/http"
+	"reflect"
+	"strings"
+)
+
+type FlagResolver interface {
+	resolveFlag(ctx context.Context, flag string, defaultValue interface{},
+		evalCtx map[string]interface{}, expectedKind reflect.Kind) InterfaceResolutionDetail
+}
+
+type ContextProvider interface {
+	GetContext() map[string]interface{}
+}
+
+var (
+	SDK_ID      = "SDK_ID_GO_PROVIDER"
+	SDK_VERSION = "0.1.8" // x-release-please-version
+)
+
+type Confidence struct {
+	parent        ContextProvider
+	uploader      EventUploader
+	contextMap    map[string]interface{}
+	Config        APIConfig
+	ResolveClient ResolveClient
+}
+
+func (e Confidence) GetContext() map[string]interface{} {
+	currentMap := e.contextMap
+	parentMap := make(map[string]interface{})
+	if e.parent != nil {
+		parentMap = e.parent.GetContext()
+	}
+	for key, value := range parentMap {
+		currentMap[key] = value
+	}
+	return currentMap
+}
+
+type ConfidenceBuilder struct {
+	confidence Confidence
+}
+
+func (e ConfidenceBuilder) SetAPIConfig(config APIConfig) ConfidenceBuilder {
+	e.confidence.Config = config
+	return e
+}
+
+func (e ConfidenceBuilder) SetResolveClient(client ResolveClient) ConfidenceBuilder {
+	e.confidence.ResolveClient = client
+	return e
+}
+
+func (e ConfidenceBuilder) Build() Confidence {
+	if e.confidence.ResolveClient == nil {
+		e.confidence.ResolveClient = HttpResolveClient{Client: &http.Client{}, Config: e.confidence.Config}
+	}
+	e.confidence.contextMap = make(map[string]interface{})
+	return e.confidence
+}
+
+func NewConfidenceBuilder() ConfidenceBuilder {
+	return ConfidenceBuilder{
+		confidence: Confidence{},
+	}
+}
+
+func (e Confidence) PutContext(key string, value interface{}) {
+	e.contextMap[key] = value
+}
+
+func (e Confidence) Track(ctx context.Context, eventName string, message map[string]interface{}) {
+	go func() {
+		currentTime := time.Now()
+		iso8601Time := currentTime.Format(time.RFC3339)
+
+		newMap := e.GetContext()
```

Comment:
I suggest we move the `getContext` call outside of the go routine.
Maybe best to actually only wrap the long running upload in it?